src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java

```diff
@@ -0,0 +1,158 @@
+    {
+        @Override
+        public <T extends Entity> List<T> matchEntities(final ICommandSender sender, final String token, final Class<? extends T> targetClass) throws CommandException
+        {
+            return EntitySelector.matchEntitiesDefault(sender, token, targetClass);
+        }
+
+        @Override
+        public boolean matchesMultiplePlayers(final String selectorStr) throws CommandException
+        {
+            return EntitySelector.matchesMultiplePlayersDefault(selectorStr);
+        }
+
+        @Override
+        public boolean isSelector(final String selectorStr)
+        {
+            return EntitySelector.isSelectorDefault(selectorStr);
+        }
+    };
+
+    static
+    {
+        for (final String prefix : ArrayUtils.toArray("@p", "@a", "@r", "@e"))
+        {
+            selectorHandlers.put(prefix, vanillaHandler);
+            registeringMods.put(prefix, "Minecraft");
+        }
+    }
+
+    /**
+     * Returns whether there is already a more specific prefix registered
+     */
+    public static boolean isShadowed(final String prefix)
+    {
+        for (final String other : registeringMods.keySet())
+            if (prefix.startsWith(other))
+                return true;
+
+        return false;
+    }
+
+    /**
+     * Returns whether there is already a less specific prefix registered
+     */
+    public static boolean isShadowing(final String prefix)
+    {
+        for (final String other : registeringMods.keySet())
+            if (other.startsWith(prefix))
+                return true;
+
+        return false;
+    }
+
+    /**
+     * Registers a new {@link SelectorHandler} for the given prefix.
+     * Warnings are emitted when {@link #isShadowing} or {@link #isShadowed} would return {@code true}
+     */
+    public static void register(final String prefix, final SelectorHandler handler)
+    {
+        final String modName = Loader.instance().activeModContainer().getName();
+
+        for (final Entry<String, String> prefixData : registeringMods.entrySet())
+        {
+            final String otherPrefix = prefixData.getKey();
+
+            if (prefix.startsWith(otherPrefix))
+                FMLLog.info("Selector prefix '%s' of mod '%s' is shadowing prefix '%s' of mod '%s'", prefix, modName, otherPrefix, prefixData.getValue());
+            else if (otherPrefix.startsWith(prefix))
+                FMLLog.info("Selector prefix '%s' of mod '%s' is shadowed by prefix '%s' of mod '%s'", prefix, modName, otherPrefix, prefixData.getValue());
+        }
+
+        selectorHandlers.put(prefix, handler);
+        registeringMods.put(prefix, modName);
+    }
+
+    /**
+     * Returns the best matching handler for the given string. Defaults to the vanilla handler if no prefix applies
+     */
+    public static SelectorHandler getHandler(final String selectorStr)
+    {
+        for (final Entry<String, SelectorHandler> handler : selectorHandlers.entrySet())
```

Comment:
This should definitely work - it may even be possible to use `selectorHandlers.subMap(selectorStr, true, selectorStr.substring(0,1), true).entrySet()`